File: Arrays/219-Contains_Duplicate_II/219-Contains_Duplicate_II.java
    // Brute Force Approach
    // Time Complexity - O(N^2)
    // Space Complexity - O(1)
    public boolean containsNearbyDuplicateBruteForce(int[] nums, int k) {
        int n = nums.length;
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                if (nums[i] == nums[j] && Math.abs(i - j) <= k) {
                    return true;
                }
            }
        }
        return false;
    }

    // Optimal Approach
    // Time Complexity - O(N)
    // Space Complexity - O(N)
    public boolean containsNearbyDuplicateOptimal(int[] nums, int k) {
        int n = nums.length;
        HashMap<Integer, Integer> map = new HashMap<>();

        for (int i = 0; i < n; i++) {
            if (map.containsKey(nums[i]) && Math.abs(i - map.get(nums[i])) <= k) {
                return true;
            } else {
                map.put(nums[i], i);
            }
        }
        return false;
    }

}

File: Arrays/219-Contains_Duplicate_II/219-Contains_Duplicate_II.cpp
//Brute Force Approach
//Time Complexity - O(N^2)
//Space Complexity - O(1)


bool containsNearbyDuplicate(vector<int>& nums, int k) {
    int n = nums.size();
    for(int i = 0; i < n; i++){
        for(int j = i+1; j < n; j++){
            if(nums[i] == nums[j] and abs(i - j) <= k) return true;
        }
    }
    return false;
}


//Optimal Approach
//Time Complexity - O(N)
//Space Complexity - O(N)

bool containsNearbyDuplicate(vector<int>& nums, int k) {
    int n = nums.size();
    unordered_map<int,int>mp;

    for(int i = 0; i < n; i++){
       if(mp.find(nums[i]) != mp.end() and abs(i - mp[nums[i]]) <= k) return true;
       else mp[nums[i]] = i;
    }
    return false;
}